### DISCUSSION TITLE
using nodemailer in nextjs can't resolve child_process

### DISCUSSION THREAD

**Discussion by steveswork** [vercel/next.js#10778] [Help]
Using nodemailer in one of the `pages/api` routes produces the following error message:
`[ error ] ./node_modules/nodemailer/lib/sendmail-transport/index.js
Module not found: Can't resolve 'child_process' in 'C:\ua-demo\node_modules\nodemailer\lib\sendmail-transport'`

With my understanding being that the next.js `pages/api` routes run in the server environment alone, it is a wonder that this error is occurring. How could this be addressed so that I can send email updates to my users?

An example has been added here at [codesandbox.io](https://codesandbox.io/s/nextjs-nodemailer-api-testshared-tmhen). I think that we need to build a copy of the program directly on a local machine to reproduce.

[Also discussed on SO](https://stackoverflow.com/questions/60496911/cant-resolve-child-process-when-using-nodemailer-in-next-js)

**Comment by Timer** [comment]
Please provide a fully reproducing example so someone may provide assistance.

**Reply by steveswork** [reply]
I will not be able to provide it. NDA aside, the code example will be voluminous to post. If it helps though, the Nodemailer package is used in a module which is a domain service that sends out update alerts to users. The `pages/api` uses this domain service.

The issue then is that when I run `npm run dev`, I get the error above. By the way,  my package.json contains this snippet: `{ "scripts": { "dev": "clear && next dev", ... }, ... }`

**Comment by steveswork** [comment]
@Timer, just wondering if you've had time to take another look at this issue? I cannot build this app without resolving this issue first. Is there another npm emailer package better suited for NextJs applications that I can use instead?

**Reply by Timer** [reply]
Like I said previously:

> Please provide a fully reproducing example so someone may provide assistance.

Without this, we're unable to provide assistance.

**Reply by steveswork** [reply]
Ok, I've updated my original post with a codesandbox link to an example.

**Comment by nicholaschiang** [comment]
Hey @Timer shouldn't Next.js automatically load server-side only modules in the `pages/api` directory? Why is this even an error?

**Comment by kmturley** [comment]
I had a similar issue. It seems that a generic package import, traverses all files underneath regardless of whether the methods are actually called:

`import { clientSideMethod } from 'mypackage'`

Actually first goes to the root index.js inside the package and calls:

```
export * from './file'; // contains server-side code
export * from './utils'; // contains client-side code only
```

Therefore the error was triggered. Solutions are:

1. Use a more specific import import { clientSideMethod } from '../../node_modules/mypackage/dist/utils'
2. Dynamically import server-side modules when needed
3. Split packages or modules into client-side vs server-side
